### PR TITLE
Staging Site: Fix transfer logic

### PR DESCRIPTION
**Pull request by kozer**
Related to https://github.com/Automattic/dotcom-forge/issues/4133

## Proposed Changes

This PR aims to fix the Staging Site card component and the different edge cases we have when we add/delete a staging site.

### Main changes
- Tried to remove all different local state and if statements. Now everything is centralized around the different states that the status can be on.
- Introduced another "status state" similar to what we have on sync. The reason for that is that I store the sync state on the production site id, and not on the staging. This is because the staging id is not available from the start, and there is a timeframe where the user can reload the page, and display the wrong state.
- I introduce a new WPCOM endpoint to fetch the lock status, in an effort to get have as much "source of truth" data from the server as possible.
- Updated `useCheckStagingSiteStatus` hook to return `''` instead of `null` if it's on a "loading" state. This helps to avoid update the "staging status" if not needed.

The problem this component has, and why it is a bit complex to get a valid state out of it, can be summarized in the next diagram:

image

**TL;DR:** There are several timeframes where we do have not a valid state. This is amplified locally due to the fact that calypso "resets" the state randomly.

This is a nonoptimal solution.

A better solution may be the following:
- Create a staging site ID, return that, and keep it in the creation phase, while in the background we do everything we need to do ( I searched about it, but couldn't find a way to do that ).
- Keep fetching that ID till the "transfer" ( add ) completes successfully. Every part of it.
- On revert, set a flag that indicates we are reverting and remove that flag when everything is completed.

Actually, we need a lock mechanism in place that keeps the state, before each action and is removed the moment that everything is stabilized.
The downside of it will be that we'll rely on the network to determine the state.

Either way, I think that the underlying changes are into the right direction, as we have a centralized place we handle the logic.

Any ideas are more than welcome.

## Testing Instructions
1. Apply D126771-code.
2. Create a new WoA site.
3. Navigate to the hosting configuration page.
4. Add a new site.
5. Try EVERYTHING you believe that can break it.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by kozer** [comment]
Try to break it before I change it's state to "review", please!

**Comment by wojtekn** [comment]
@kozer, thanks for working on this one. I didn't review the whole code, but your note above the chart made me curious if this issue is valid only for development environments:

> NOTE: Cache can be deleted anytime ( develop simulates that )

I see we have some mechanism in place, which is using `wasCreating` to display the 'Staging site added.' message only once. Does that issue occur only, or primarily, in a development environment where the Redux store cache is cleaned randomly? If yes, is it worth introducing a complex code change to fix that?

**Comment by kozer** [comment]
> I see we have some mechanism in place, which is using `wasCreating` to display the 'Staging site added.' message only once. Does that issue occur only, or primarily, in a development environment where the Redux store cache is cleaned randomly? If yes, is it worth introducing a complex code change to fix that?

Yes, @wojtekn, it seems it is introduced only then, from my experience. This is something I mentioned [here](p1698249798062519-slack-C04J3B7JE92), but I continue working on that based on @danielbachhuber recommendation.

I don't know if this is more complex than it was before. If you remove the boilerplate code of the reducer, the main changes are in the staging-site component and are not that many.
Maybe it seems more complex than it is due to the reducer introduction, and some corner cases.

However, I find both implementations too complex for what we want to achieve here.
We need to find a way to obtain that information from the backend, without having those "missing state gaps"

File: client/state/staging-site/selectors/get-staging-site-info.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import type { AppState } from 'calypso/types';
+
+import 'calypso/state/staging-site/init';
+
+const emptyData = {};
+
+export const getStagingSiteInfo = ( state: AppState, siteId: number | null ) =>
+	siteId ? state?.stagingSite?.[ siteId ] ?? emptyData : emptyData;
```
Comment (sejas) [inline client/state/staging-site/selectors/get-staging-site-info.ts]: Could we have an early return, so this expression is more clear?

File: client/state/staging-site/selectors/get-staging-site-status.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Returns status info for transfer
+ * @param {Object} state global app state
+ * @param {number} siteId requested site for transfer info
+ * @returns {string|null} transferStatusType if available else ''
+ */
+export function getStagingSiteStatus( state: AppState, siteId: number | null ): string | null {
+	return getStagingSiteStatusData( getStagingSiteInfo( state, siteId ) );
+}
```
Comment (sejas) [inline client/state/staging-site/selectors/get-staging-site-status.ts:23]: What's the benefit of having three functions that are almost identical?

`getStagingSiteStatusData` vs `getStagingSiteStatus` vs `getStagingSiteInfo` ?

**Review by sejas — COMMENTED** [review]
Using redux seems reasonable to fix the current issue. Another alternative would be making all the fetch requests (siteInfo, getAutomatedTransferStatus, and lock) in the same react query hook. So we get all the information we need at once producing only one render when we have enough information, until that moment we would be "loading" with the placeholder skeleton.
I added some questions to the code. I didn't test it yet.

File: client/state/staging-site/selectors/get-staging-site-status.ts (hunk 2 above)
Comment (kozer) [reply client/state/staging-site/selectors/get-staging-site-status.ts:23]: - `getStagingSiteInfo`: fetches the data for the specific site, as the data are in the following format:
```
{
 1212312312: {
   "status": '...'
}
```
-`getStagingSiteStatusData` and `getStagingSiteStatus`: I'm just following the pattern we use on other reducers here. One that is used to other reducers, will understand with a first glance what is happening, without even read the code.

**Comment by kozer** [comment]
> So we get all the information we need at once producing only one render when we have enough information, until that moment we would be "loading" with the placeholder skeleton.

That's a good idea.
But `getAutomatedTransferStatus` needs to be polled. Can you give a small example to understand better what you have in mind?

**Comment by sejas** [comment]
> But getAutomatedTransferStatus needs to be polled. Can you give a small example to understand better what you have in mind?

Yes, what I mean is that we can make the first request with `Promise.all`, and then we would need to check the status in an active wait. Happy to pair on this.

File: client/state/staging-site/constants.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export enum StagingSiteStatus {
+	UNSET = '',
+	NONE = 'none',
+	REVERTING = 'reverting',
+	INITIATE_TRANSFERRING = 'initiate transferring',
+	INITIATE_REVERTING = 'initiate reverting',
+	TRANSFERRING = 'transferring',
+	COMPLETE = 'complete',
+	REVERTED = 'reverted',
+}
```
Comment (sejas) [inline client/state/staging-site/constants.ts]: What do you think if we reuse the existing transfer status?

https://github.com/Automattic/wp-calypso/blob/6704bd781443cc4a60f10949844a573dc6dc20f8/client/state/automated-transfer/constants.ts#L1-L28

You are already using them on `client/state/automated-transfer/constants.ts:30`

File: client/my-sites/hosting/staging-site-card/index.js [hunk 4]
```diff
@@ -81,26 +93,58 @@ export const StagingSiteCard = ( {
...
+	const transferStatus = useCheckStagingSiteStatus( stagingSite.id, hasCompletedInitialLoading );
+
+	const isStagingSiteTransferComplete = useMemo( () => {
+		return (
+			stagingSiteStatus === StagingSiteStatus.COMPLETE ||
+			stagingSiteStatus === StagingSiteStatus.REVERTED ||
+			stagingSiteStatus === StagingSiteStatus.NONE ||
+			stagingSiteStatus === StagingSiteStatus.UNSET
+		);
+	}, [ stagingSiteStatus ] );
```
Comment (sejas) [inline client/my-sites/hosting/staging-site-card/index.js]: We could move this code into a new selector. No need to useMemo, since it's not a heavy computation.

File: client/my-sites/hosting/staging-site-card/index.js [hunk 5]
```diff
@@ -116,27 +160,34 @@ export const StagingSiteCard = ( {
...
-			} );
+		if ( stagingSiteStatus === StagingSiteStatus.COMPLETE ) {
+			queryClient.invalidateQueries( [ USE_SITE_EXCERPTS_QUERY_KEY ] );
+			dispatch( setStagingSiteStatus( siteId, StagingSiteStatus.NONE ) );
 			dispatch(
 				successNotice( __( 'Staging site added.' ), { id: stagingSiteAddSuccessNoticeId } )
 			);
 		}
-	}, [ dispatch, queryClient, __, isStagingSiteTransferComplete, wasCreating ] );
+	}, [ dispatch, queryClient, __, isStagingSiteTransferComplete, siteId, stagingSiteStatus ] );
```
Comment (sejas) [inline client/my-sites/hosting/staging-site-card/index.js]: I think we can remove `isStagingSiteTransferComplete` form the dependencies array:

```suggestion
	}, [ dispatch, queryClient, __, siteId, stagingSiteStatus ] );
```

**Review by sejas — APPROVED** [review]
Tested deeply and couldn't break the transfers. I didn't get any duplicated noticies.

Added a few comments like reusing the transfer statuses instead of creating new ones.

Thanks for adding the tests! 🙌 .

https://github.com/Automattic/wp-calypso/assets/779993/20dd3329-418e-4a73-aa9c-640777e8db58

**Comment by danielbachhuber** [comment]
@kozer Can we use this approach to address https://github.com/Automattic/wp-calypso/issues/72371 too?

**Comment by kozer** [comment]
> @kozer Can we use this approach to address https://github.com/Automattic/wp-calypso/issues/72371 too?

@danielbachhuber , I guess we can, but I need to dig into this more now that I understand where the problem may have originated.
Let's handle that separately, as this one is already too big. I can take a  look after this one gets merged.

File: client/state/staging-site/constants.ts (hunk 3 above)
Comment (kozer) [reply client/state/staging-site/constants.ts]: How do you think we should reuse that? Actually what it's happening now is a mapping from more to less states, in order to have better control of what's happening.
Thanks for your comment!

Comment (sejas) [reply client/state/staging-site/constants.ts]: @kozer, Thanks for mapping the existing states https://github.com/Automattic/wp-calypso/pull/83545/commits/b90185245e949c47222e489ea836e70272efb127

This allows us to differentiate states coming from the network and synthetic states, like `initiate transferring`.

All good on my side to merge 👌

**Comment by tyxla** [comment]
It appears this PR introduced usage of React Query `onSuccess` and `onError.` Those are removed in v5, and as we've been trying to update to v5 (see #84413 and #84338), in #86149 I've suggested a minor refactor that could use your review.
